### PR TITLE
Minor: Add Documentation and Examples to `TableReference`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Related to https://github.com/apache/arrow-datafusion/pull/5615

# Rationale for this change
While working on https://github.com/apache/arrow-datafusion/pull/5615 I realized how complicated the TableReference has become so I felt more documentataion was in order

# What changes are included in this PR?

Add Documentation and Examples to `TableReference`

# Are these changes tested?

Yes, doc tests are tested

# Are there any user-facing changes?

More docs

### CODE REVIEW

File: datafusion/common/src/table_reference.rs [hunk 1]
```diff
@@ -35,7 +35,38 @@ impl<'a> std::fmt::Display for ResolvedTableReference<'a> {
     }
 }
 
-/// Represents a path to a table that may require further resolution
+/// [`TableReference`]s represent a multi part identifier (path) to a
+/// table that may require further resolution.
+///
+/// # Creating [`TableReference`]
+///
+/// When converting strings to [`TableReference`]s, the string is
```
Comment (alamb) [inline datafusion/common/src/table_reference.rs:43]: The SQL parsing behavior added in https://github.com/apache/arrow-datafusion/pull/5343 is somewhat surprising so I think it should be well documented

**Comment by alamb** [comment]
FYI @Jefffrey

File: datafusion/common/src/table_reference.rs [hunk 2]
```diff
@@ -80,14 +121,18 @@ impl std::fmt::Display for TableReference<'_> {
...
+    /// all -- so "Foo.Bar" stays as a reference to the table named
+    /// "Foo.Bar" (rather than "foo"."bar")
     pub fn bare(table: impl Into<Cow<'a, str>>) -> TableReference<'a> {
         TableReference::Bare {
             table: table.into(),
         }
     }
 
-    /// Convenience method for creating a `Partial` variant of `TableReference`
+    /// Convenience method for creating a [`TableReference::Partial`]
```
Comment (Jefffrey) [inline datafusion/common/src/table_reference.rs:135]: maybe also mention these do no parsing either for their parts? ditto for full

File: datafusion/common/src/table_reference.rs [hunk 3]
```diff
@@ -189,7 +234,8 @@ impl<'a> TableReference<'a> {
         }
     }
 
-    /// Converts directly into an [`OwnedTableReference`]
+    /// Converts directly into an [`OwnedTableReference`] by copying
```
Comment (Jefffrey) [inline datafusion/common/src/table_reference.rs:237]: copying -> cloning maybe?

not sure how explicit need to be about cloning in documentation

File: datafusion/common/src/table_reference.rs [hunk 4]
```diff
@@ -260,7 +310,7 @@ impl<'a> TableReference<'a> {
     }
 }
 
-/// Parse a `String` into a OwnedTableReference
+/// Parse a `String` into a OwnedTableReference as a SQL identifier.
```
Comment (Jefffrey) [inline datafusion/common/src/table_reference.rs:313]: `... as a multipart SQL identifier.`

File: datafusion/common/src/table_reference.rs [hunk 5]
```diff
@@ -231,14 +287,8 @@ impl<'a> TableReference<'a> {
...
-    /// Forms a [`TableReference`] by attempting to parse `s` as a multipart identifier,
-    /// failing that then taking the entire unnormalized input as the identifier itself.
-    ///
-    /// Will normalize (convert to lowercase) any unquoted identifiers.
-    ///
-    /// e.g. `Foo` will be parsed as `foo`, and `"Foo"".bar"` will be parsed as
-    /// `Foo".bar` (note the preserved case and requiring two double quotes to represent
-    /// a single double quote in the identifier)
+    /// Forms a [`TableReference`] by parsing `s` as a multipart
+    /// identifier. See docs on [`TableReference`] for more details.
```
Comment (Jefffrey) [inline datafusion/common/src/table_reference.rs:291]: `... as a multipart SQL identifier.`

**Review by Jefffrey — APPROVED** [review]
thanks for this, docs look good, just a few minor suggestions for being a bit more explicit about details

File: datafusion/common/src/table_reference.rs (hunk 2 above)
Comment (alamb) [reply datafusion/common/src/table_reference.rs]: 👍  in fd68dc057

File: datafusion/common/src/table_reference.rs [hunk 6]
```diff
@@ -80,14 +121,20 @@ impl std::fmt::Display for TableReference<'_> {
...
     pub fn bare(table: impl Into<Cow<'a, str>>) -> TableReference<'a> {
         TableReference::Bare {
             table: table.into(),
         }
     }
 
-    /// Convenience method for creating a `Partial` variant of `TableReference`
+    /// Convenience method for creating a [`TableReference::Partial`].
+    ///
+    /// As described on [`TableReference`] this does parsing at all.
```
Comment (Jefffrey) [inline datafusion/common/src/table_reference.rs:137]: `... this does *NO* parsing at all.`

ditto for `full`

Comment (alamb) [reply datafusion/common/src/table_reference.rs:142]:
```suggestion
    /// As described on [`TableReference`] this does *NO* parsing at all.
```

Comment (alamb) [reply datafusion/common/src/table_reference.rs:142]: Excellent call -- thank you

File: datafusion/common/src/table_reference.rs [hunk 7]
```diff
@@ -103,7 +150,9 @@ impl<'a> TableReference<'a> {
         }
     }
 
-    /// Convenience method for creating a `Full` variant of `TableReference`
+    /// Convenience method for creating a [`TableReference::Full`]
+    ///
+    /// As described on [`TableReference`] this does parsing at all.
```
Comment (alamb) [inline datafusion/common/src/table_reference.rs:155]:
```suggestion
    /// As described on [`TableReference`] this does *NO* parsing at all.
```

**Pull request merged**